File: cpp/day2/day2p2.cpp
#include <iostream>
#include <fstream>
#include <vector>

//report class
class report {
  public:
    // create report, takes info from file and loads into report_list
    report(std::ifstream& file) {
      int tmp;
      char c = ' ';
      //while peek is 'space', newline or eof will break loop
      while (c == ' ') {
        //load int into array
        file >> tmp;
        report_list.insert(report_list.end(), tmp);

        //load next peek
        c = file.peek();
      }
      dampen = true;
    }
    //checks if the pair in report within safe bounds
    bool is_safe(int& diff) {
      //inc
      if (type) {
        return (diff >= -3 && diff < 0);
      }
      //dec
      else {
        return (diff <= 3 && diff > 0);
      }
    }

    bool run_report() {
      for (int i = 1; i < report_list.size(); i++) {
        int diff = report_list[i - 1] - report_list[i];
        //check if the pair in the report is safe
        if (!is_safe(diff)) {
          if(dampen) {
            if (i + 1 == report_list.size()) {
              report_list.erase(report_list.begin() + i);
              dampen = false;
              return run_report();
            }
            diff = report_list[i - 1] - report_list[i + 1];

            if (is_safe(diff)) {
              report_list.erase(report_list.begin() + i);
              dampen = false;
              return run_report();
            }
            else {
              report_list.erase(report_list.begin() + i - 1);
              dampen = false;
              return run_report();
            }
          }
          else {
            return false;
          }
        }
      }
      return true;
    }
    
    //return true to skip this report, false keeps while loop going and sets type to increasing or decreasing
    bool check_type() {
      int diff[6];
      //check first diff
      diff[0] = report_list[0] - report_list[1];
      diff[1] = report_list[0] - report_list[2];
      diff[2] = report_list[0] - report_list[3];
      diff[3] = report_list[1] - report_list[2];
      diff[4] = report_list[1] - report_list[3];
      diff[5] = report_list[2] - report_list[3];

      //if diff == 0 for 1st and 2nd element then skip this report
      if (diff[0] == 0 && diff[1] == 0) {
        return true;
      }

      int count = 0;
      //loop over diff array(6 being the len of diff)
      for (int i = 0; i < 6; i++) {
        if (diff[i] < 0) {
          count++;
        }
        else if (diff[i] > 0) {
          count--;
        }
      }
      
      //if count is bigger than 0 then it is increasing, if less then decreasing
      if (count > 0) {
        type = true;
      }
      else if (count < 0) {
        type = false;
      }
      return false;
    }



  private:
    std::vector<int> report_list;
    // if type is true then the report is increasing, false means decreasing
    bool type;
    // if dampen is true then it has not be used and can be used to skip a number in a report
    bool dampen;
};


int main() {
  //open file
  std::ifstream file("day2.txt");

  int count = 0;
  bool type = true;

  //iterate over file reports
  while (true) {
    //create report
    report file_report(file); 

    if (file.eof()) {
      break;
    }

    //skip this report if needed
    if (file_report.check_type()) {
      continue;
    }

    
    //run report to see if it is safe, if true then add to count
    if(file_report.run_report()) {
      count++;
    }
  }

  std::cout << "The amount of safe reports is: " << count << '\n';
  return 0;
}

File: Lua/day2/day2p2.lua
-- Given lines of ints
-- Find out how many of the lines are safe.
-- A line is safe if the ints are increasing or decreasing by increments of 1-3.
-- A line is not safe if it does both increasing and decreasing or increments more than 3 or less than 1(0)

--problem 2 change: check if one number in the report can be removed to make the report safe

-- func def
local function strToTable(s)
  -- input: string (s) of numbers separated by one space each.
  -- output: a table (t) consisting of ints from the input string.

  local t = {}
  local num = ""

  for c in s:gmatch(".") do
    if c == " " then
      table.insert(t, tonumber(num))
      num = ""
    else
      num = num .. c
    end
  end
  table.insert(t, tonumber(num))
  return t
end

--[[

main

]]

local file = io.open("day2test.txt", "r")

if file == nil then
  print("Can not open file")
  return
end

-- vars
local report = {}
local diff
-- skip is special, 0 is no skips have happend, 1 is skip has happend and next loop should skip the number, 2 means there has already been a skip 
local skip
local case
local count = 0

-- loop over each line in file to find the 'safe' reports
for line in file:lines("l") do

  -- turn line into table
  report = strToTable(line)
  diff = report[1] - report[2]

  -- reset 
  case = true
  skip = 0

  if diff == 0 then
    diff = report[2] - report[3]
  end

  -- decreasing
  if diff > 0 then
    for i in #report do
      if i == 1 then
        goto next
      end
      if skip == 0 or skip == 2 then
        diff = report[i - 1] - report[i]
      elseif skip == 1 then
        diff = report[i - 2] - report[i]
        skip = 2
      end

      -- check case
      if not (diff > 0 and diff <= 3 and diff ~= 0) then
        if skip ~= 2 then
          skip = 1
        else
          case = false
          break
        end
      end
      ::next::
    end

    -- check if loop completed properly
    if case == true then
      count = count + 1
    end

  -- increasing
  elseif diff < 0 then
    for i in #report do
      if i == 1 then
        goto next
      if skip == 0 or skip == 2 then
        diff = report[i - 1] - report[i]
      elseif skip == 1 then
        diff = report[i - 2] - report[i]
        skip = 2
      end

      if not (diff < 0 and diff >= -3 and diff ~= 0) then
        if skip ~= 2 then
          skip = 1
        else
          case = false
          break
        end
      end
      ::next::
    end

    -- check loop
    if case == true then
      count = count + 1
    end
  end
end
end
